app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadViewModel.kt

```diff
@@ -74,31 +76,38 @@ class ThreadViewModel @Inject constructor(
         AccountUtils.currentMailboxId,
     ).map { it.obj }.asLiveData(ioCoroutineContext)
 
-    val threadLive = liveData(ioCoroutineContext) {
-        emitSource(threadController.getThreadAsync(threadUid).map { it.obj }.asLiveData())
+    fun reassignThreadLive(threadUid: String) {
+        threadLiveJob?.cancel()
+        threadLiveJob = viewModelScope.launch(ioCoroutineContext) {
+            threadController.getThreadAsync(threadUid).map { it.obj }.collect(threadLive::postValue)
+        }
     }
 
-    val messagesLive = liveData(ioCoroutineContext) {
+    fun reassignMessagesLive(threadUid: String) {
+        messagesLiveJob?.cancel()
+        messagesLiveJob = viewModelScope.launch(ioCoroutineContext) {
 
-        suspend fun splitBody(message: Message): Message = withContext(ioDispatcher) {
-            if (message.body == null) return@withContext message
+            suspend fun splitBody(message: Message): Message = withContext(ioDispatcher) {
+                if (message.body == null) return@withContext message
 
-            return@withContext message.apply {
-                body?.let {
-                    val isNotAlreadySplit = !splitBodies.contains(message.uid)
-                    if (isNotAlreadySplit) splitBodies[message.uid] = MessageBodyUtils.splitContentAndQuote(it)
-                    splitBody = splitBodies[message.uid]
+                return@withContext message.apply {
+                    body?.let {
+                        val isNotAlreadySplit = !splitBodies.contains(message.uid)
+                        if (isNotAlreadySplit) splitBodies[message.uid] = MessageBodyUtils.splitContentAndQuote(it)
+                        splitBody = splitBodies[message.uid]
+                    }
                 }
             }
-        }
 
-        messageController.getSortedAndNotDeletedMessagesAsync(threadUid)
-            ?.map { results -> results.list.map { splitBody(it) } }
-            ?.asLiveData()
-            ?.let { emitSource(it) }
+            splitBodies.clear()
+
+            messageController.getSortedAndNotDeletedMessagesAsync(threadUid)
+                ?.map { results -> results.list.map { splitBody(it) } }
+                ?.collect(messagesLive::postValue)
+        }
```

Comment:
This method is a bit complicated to understand. Maybe something like this? 

```kotlin
fun reassignMessagesLive(threadUid: String) {
        messagesLiveJob?.cancel()
        messagesLiveJob = viewModelScope.launch(ioCoroutineContext) {

            suspend fun splitBody(message: Message): Message = withContext(ioDispatcher) {
                val splitBodies = mutableMapOf<String, SplitBody>()

                if (message.body == null) return@withContext message

                message.apply {
                    body?.let {
                        val isNotAlreadySplit = !splitBodies.contains(message.uid)
                        if (isNotAlreadySplit) splitBodies[message.uid] = MessageBodyUtils.splitContentAndQuote(it)
                        splitBody = splitBodies[message.uid]
                    }
                }

                return@withContext message
            }

            messageController.getSortedAndNotDeletedMessagesAsync(threadUid)
                ?.map { results -> results.list.map { splitBody(it) } }
                ?.collect(messagesLive::postValue)
        }
    }
```

Also, you don't need that splitBodies variable to be a class variable. It could be within the inner function  ? Like in my suggestion ?

Reply:
True, `splitBodies` doesn't need to be a class variable, but we can't put it in `fun splitBody()`.
If we do that, it will reset each time `splitBody()` is called, defeating its purpose.

Instead, we can put it in `reassignMessagesLive ()`, just after launching the coroutine.